test/unit.js

```diff
@@ -70,7 +70,6 @@ describe('SQL', () => {
       assert.equal(statement.text, 'SELECT * FROM table WHERE column = $1 ORDER BY other_column')
       assert.strictEqual(statement.query, 'SELECT * FROM table WHERE column = $1 ORDER BY other_column')
       assert.strictEqual(statement.values, undefined)
-      assert.strictEqual('values' in statement, false)
```

Comment:
Why did you remove this assertion? I vaguely remember that this was added intentionally because one of the DB drivers used a `for in` loop to iterate over options. So it is critical that the property is enumerable. Not passing this assertion would mean a regression and break compatibility. This is why there is an `Object.defineProperty()` call in the version on master.

Reply:
like I've said, adding and removing properties to an object makes the object slow.
If you understand the changes I've made, you'll realize properties are getters and returned accordingly to the `bound` state.
If this breaks anywhere, it should be stated where, which version, and why.

I can move these two outside as non enumerable, if that solves the `for in` issue, but that won't solve `in` ... which is a different story.

So, in few words, are you asking for non enumerables, or for runtime deleted properties that makes this whole optimization useless?

Reply:
I understand the changes you made. The problem is the integration tests cannot detect whether a query was actually executed as a prepared statement or not, so you'll have to verify that manually.
It seems like the only constraint though is that the property is enumerable: https://sourcegraph.com/github.com/mysqljs/mysql@2a9239871fb24515485476e7c4d1deb941f525e9/-/blob/lib/Connection.js#L39-41
I couldn't find another gotcha from reading the DB driver code. Still, it should be tested.

Reply:
That deriver overwrites `.values` regardless of the `for/in` https://sourcegraph.com/github.com/mysqljs/mysql@2a9239871fb24515485476e7c4d1deb941f525e9/-/blob/lib/Connection.js#L46 and in your current code, you set the `values` as writable, enumerable, and configurable, in the constructor.

So, basically, nothing changes at all for that driver: it was setting `options.values = sql.values` regardless even before these changes.

Reply:
TL;DR if mysql/I driver never uses `useBind(true)` this change won't have any undesired side-effect. If it does, I don't know how that driver could work.

If I make both `values` and `bind` non enumerable, the `options` will never have values in case the passed one is a callback, and I don't know how that would behave.

I understand your concerns though, but I'd like to see a practical failure with current object.

Reply:
Did you test it at runtime (with the other drivers too)?

Reply:
nope, but I guess that'd be an improvement for this library in general ... I mean, having tests to validate noting breaks, with all supported drivers.

Reply:
Just to clarify, I wasn't planing to push those tests so, I wonder if we're moving forward with this PR ('cause I have more coming to improve `append` etc)